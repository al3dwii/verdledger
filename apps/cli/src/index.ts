#!/usr/bin/env node
import { Octokit } from '@octokit/rest';
import prompts from 'prompts';
import ora from 'ora';
import yaml from 'js-yaml';
import chalk from 'chalk';

(async () => {
  console.log(chalk.greenBright('\nVerdLedger one-liner installer\n'));

  const { repo, token, apiKey } = await prompts([
    { type: 'text', name: 'repo', message: 'GitHub repo (owner/name)' },
    { type: 'password', name: 'token', message: 'GitHub PAT (repo scope)' },
    { type: 'text', name: 'apiKey', message: 'VerdLedger API key (dashboard)' }
  ]);
  const [owner, name] = String(repo).split('/');
  const gh = new Octokit({ auth: token });

  const spinner = ora('Creating GitHub secretâ€¦').start();
  const { data: { key, key_id } } = await gh.actions.getRepoPublicKey({ owner, repo: name });
  const sodium = await import('tweetsodium');
  const encrypted = sodium.sodium_seal(Buffer.from(apiKey), Buffer.from(key, 'base64'));
  await gh.actions.createOrUpdateRepoSecret({
    owner,
    repo: name,
    secret_name: 'VERDLEDGER_KEY',
    encrypted_value: encrypted,
    key_id
  });
  spinner.succeed('Secret added');

  const { data: repoData } = await gh.repos.get({ owner, repo: name });
  const branch = repoData.default_branch;

  const wf = {
    name: 'VerdLedger',
    on: { pull_request: null },
    jobs: {
      advise: {
        'runs-on': 'ubuntu-latest',
        steps: [
          { uses: 'actions/checkout@v4' },
          { uses: 'hashicorp/setup-terraform@v3' },
          { run: 'terraform init' },
          { run: 'terraform plan -out=plan' },
          { run: 'terraform show -json plan > plan.json' },
          {
            uses: 'verdledger/iac-advisor-action@v0.1.0',
            with: {
              'plan-json': 'plan.json',
              'api-key': '${{ secrets.VERDLEDGER_KEY }}',
              'api-url': 'https://api.verdledger.dev'
            }
          }
        ]
      }
    }
  } as const;

  const wfPath = '.github/workflows/verdledger.yml';
  const branchName = 'verdledger/init';
  spinner.start('Pushing workflow branchâ€¦');
  const contentBase64 = Buffer.from(yaml.dump(wf)).toString('base64');
  await gh.repos.createOrUpdateFileContents({
    owner,
    repo: name,
    path: wfPath,
    message: 'Add VerdLedger Action',
    content: contentBase64,
    branch: branchName
  });
  await gh.pulls.create({
    owner,
    repo: name,
    head: branchName,
    base: branch,
    title: 'Add VerdLedger IaC Advisor',
    body: 'ðŸš€ One-click commit-time carbon & cost suggestions'
  });
  spinner.succeed(`PR opened on branch ${branchName}`);

  console.log(chalk.green('\nâœ… All done â€“ open the pull request, merge, and enjoy!\n'));
})();
