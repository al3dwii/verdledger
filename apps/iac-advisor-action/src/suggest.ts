import fetch from 'node-fetch';
import type { Resource } from './plan';

interface SkuRow {
  cloud    : string;
  region   : string;
  sku      : string;
  watts    : number;
  usd_hour : number;
}

export type Suggestion = Resource & {
  altSku   : string;
  deltaKg  : number;
  deltaUsd : number;
};

/* ---------- helper tables & functions --------------------------- */
const SIZE_DOWN: Record<string, string | undefined> = {
  nano: '-', micro: 'nano', small: 'micro', medium: 'small',
  large: 'medium', xlarge: 'large', '2xlarge': 'xlarge',
};

function downsize(sku: string): string {
  return sku.replace(
    /(nano|micro|small|medium|large|xlarge|2xlarge)/,
    m => SIZE_DOWN[m] ?? m,
  );
}

function lookup(table: Map<string, SkuRow>, provider: string, region: string, sku: string): SkuRow | undefined {
  return table.get(`${provider}/${region}/${sku}`);
}

function lookupAnyRegion(rows: SkuRow[], provider: string, sku: string): SkuRow | undefined {
  return rows.find(r => r.cloud === provider && r.sku === sku);
}

/* ---------- main suggest() -------------------------------------- */
export async function suggest(
  resources: Resource[],
  api: string,
  apiKey: string,
): Promise<Suggestion[]> {
  const headers = apiKey ? { Authorization: `Bearer ${apiKey}` } : undefined;
  const skus: SkuRow[] = await fetch(`${api}/v1/skus`, { headers }).then(r => r.json());

  const table = new Map(
    skus.map(s => [`${s.cloud}/${s.region}/${s.sku}`, s]),
  );

  return resources
    .map(resource => {
      const { provider, region, sku } = resource;
      const base = lookup(table, provider, region, sku);

      if (!base) return null;                         // unknown baseline → skip

      /* 1️⃣  Try AMD equivalent first (t3 → t3a) */
      let altSku = sku.replace(/^t3\./, 't3a.');
      let alt    = lookup(table, provider, region, altSku)
                ?? lookupAnyRegion(skus, provider, altSku);

      /* 2️⃣  If no AMD alt, downsize (“medium”→“small”, etc.) */
      if (!alt) {
        altSku = downsize(sku);
        alt    = lookup(table, provider, region, altSku)
              ?? lookupAnyRegion(skus, provider, altSku);
      }

      /* 3️⃣  Still nothing? bail out */
      if (!alt || alt === base) return null;

      // const deltaKg  = ((base.watts - alt.watts) / 1000) * 0.7;   // kg CO₂/month
      const deltaKg  = ((base.watts - alt.watts) * 730 / 1000) * 0.7;
      const deltaUsd = (base.usd_hour - alt.usd_hour) * 730;      // USD/month

      return { ...resource, altSku, deltaKg, deltaUsd };
    })
    .filter((s): s is Suggestion => !!s && s.deltaKg > 0.01);      // keep ≥ 10 g savings
}


// // import fetch from 'node-fetch';
// // import type { Resource } from './plan';

// // export type Suggestion = Resource & { altSku: string; deltaKg: number; deltaUsd: number };

// // export async function suggest(resources: Resource[], api: string): Promise<Suggestion[]> {
// //   const skus = await fetch(`${api}/v1/skus`).then(r => r.json());
// //   const table = new Map(skus.map((s: any) => [`${s.cloud}/${s.region}/${s.sku}`, s]));

// //   return resources.map(r => {
// //     const base = table.get(`${r.provider}/${r.region}/${r.sku}`)!;
// //     // naive greener choice: same family, "small" instead of "medium"
// //     const altSku = r.sku.replace(/medium/,'small');
// //     const alt    = table.get(`${r.provider}/${r.region}/${altSku}`) ?? base;

// //     const deltaKg  = ((base.watts - alt.watts) / 1000) * 0.7; // 0.7 kg/kWh default
// //     const deltaUsd = (base.usd_hour - alt.usd_hour) * 730;    // monthly

// //     return { ...r, altSku, deltaKg, deltaUsd };
// //   }).filter(s => s.deltaKg > 0.1);   // only if meaningful
// // }

// import fetch from 'node-fetch';
// import type { Resource } from './plan';

// interface SkuRow {
//   cloud: string;
//   region: string;
//   sku: string;
//   watts: number;
//   usd_hour: number;
// }

// export type Suggestion = Resource & {
//   altSku:   string;
//   deltaKg:  number;
//   deltaUsd: number;
// };

// export async function suggest(
//   resources: Resource[],
//   api: string
// ): Promise<Suggestion[]> {
//   // --- fix: tell TS the JSON shape -------------------------------▼
//   const skus = await fetch(`${api}/v1/skus`)
//     .then(res => res.json() as Promise<SkuRow[]>);

//   const table = new Map(
//     skus.map(s => [`${s.cloud}/${s.region}/${s.sku}`, s])
//   );

//   return resources
//     .map(r => {
//       const base = table.get(`${r.provider}/${r.region}/${r.sku}`)!;

//       const altSku = r.sku.replace(/medium/, 'small');
//       const alt    = table.get(`${r.provider}/${r.region}/${altSku}`) ?? base;

//       const deltaKg  = ((base.watts - alt.watts) / 1000) * 0.7;
//       const deltaUsd = (base.usd_hour - alt.usd_hour) * 730;

//       return { ...r, altSku, deltaKg, deltaUsd };
//     })
//     .filter(s => s.deltaKg > 0.1);
// }
