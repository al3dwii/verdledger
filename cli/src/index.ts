#!/usr/bin/env node
import { Command } from 'commander';
import { createClient } from '@supabase/supabase-js';
import fs from 'fs';

const program = new Command();
program.name('verd').description('VerdLedger CLI').version('0.1.0');

program
  .command('push <file>')
  .description('Push a local plan result to the ledger')
  .action(async (file) => {
    const content = fs.readFileSync(file, 'utf8');
    const data = JSON.parse(content);
    const supabase = createClient(process.env.SUPABASE_URL || '', process.env.SUPABASE_SERVICE_KEY || '');
    await supabase.from('savings_event').insert(data);
    console.log('Pushed events to ledger');
  });

program.parse(process.argv);
