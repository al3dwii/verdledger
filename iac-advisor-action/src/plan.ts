import fs from 'fs';

export type Resource = {
  provider: string;            // "aws"
  region:   string;            // "eu-central-1"
  sku:      string;            // "t3.medium"
};

export function parsePlan(path: string): Resource[] {
  const plan = JSON.parse(fs.readFileSync(path,'utf8'));
  const resources: Resource[] = [];

  plan.resource_changes?.forEach((rc: any) => {
    if (rc.change.actions.includes('create')) {
      const provReg = rc.address.split('.')[0];
      const provider = provReg.split('_')[0];
      resources.push({
        provider,
        region:  rc.change.after.availability_zone.replace(/[a-z]$/,''),
        sku:     rc.change.after.instance_type
      });
    }
  });
  return resources;
}
