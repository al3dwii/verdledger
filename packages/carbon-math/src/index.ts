export interface Plan {
  resources: Array<{name: string}>;
}

export function calculateFootprint(plan: Plan) {
  return {
    resourceCount: plan.resources.length,
  };
}
