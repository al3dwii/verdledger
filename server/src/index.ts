import express from 'express';
import { createClient } from '@supabase/supabase-js';

const PORT = process.env.PORT || 3000;
const SUPABASE_URL = process.env.SUPABASE_URL || '';
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || '';

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

const app = express();
app.use(express.json());

app.get('/v1/skus', async (_req, res) => {
  const { data, error } = await supabase.from('sku_catalogue').select('*');
  if (error) return res.status(500).json({ error: error.message });
  res.json(data);
});

app.get('/v1/events', async (_req, res) => {
  const { data, error } = await supabase.from('savings_event').select('*');
  if (error) return res.status(500).json({ error: error.message });
  res.json(data);
});

app.get('/v1/summary', async (_req, res) => {
  const { data, error } = await supabase
    .from('savings_event')
    .select('org_id, sum(kwh) as kwh, sum(kg) as kg, sum(usd) as usd')
    .group('org_id');
  if (error) return res.status(500).json({ error: error.message });
  res.json(data);
});

app.listen(PORT, () => {
  console.log(`VerdLedger API listening on port ${PORT}`);
});
