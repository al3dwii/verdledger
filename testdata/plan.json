{
  "format_version": "1.1",
  "terraform_version": "1.7.5",
  "resource_changes": [
    {
      "address": "aws_instance.app",
      "mode": "managed",
      "type": "aws_instance",
      "name": "app",
      "provider_name": "registry.terraform.io/hashicorp/aws",
      "change": {
        "actions": ["create"],
        "after": {
          "availability_zone": "us-west-2a",
          "instance_type": "t3.medium"
        }
      }
    }
  ]
}
