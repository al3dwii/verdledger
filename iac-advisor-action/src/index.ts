import * as core from '@actions/core';
import * as github from '@actions/github';
import { parsePlan } from './plan';
import { suggest }   from './suggest';
import fetch from 'node-fetch';

async function run() {
  try {
    const api   = core.getInput('api-url');
    const key   = core.getInput('api-key');
    const plan  = core.getInput('plan-json');

    const resources = parsePlan(plan);
    const sugg      = await suggest(resources, api);

    if (!sugg.length) {
      core.info('No greener alternatives found.');
      return;
    }

    const markdown = [
      `### \u2618\uFE0F VerdLedger suggestions`,
      '| Current | Greener | \u0394 kg CO\u2082 | \u0394 Cost |',
      '|---------|---------|---------|-------|',
      ...sugg.map(s => `| \`${s.sku}\` | \`${s.altSku}\` | **${s.deltaKg.toFixed(2)}** | **$${s.deltaUsd.toFixed(2)}** |`)
    ].join('\n');

    const token = process.env.GITHUB_TOKEN!;
    const octo  = github.getOctokit(token);
    const { owner, repo, number } = github.context.issue;

    await octo.rest.issues.createComment({ owner, repo, issue_number: number, body: markdown });

    // Listen for "verdledger:apply" label on later job
    core.setOutput('suggestions', JSON.stringify(sugg));

  } catch (err) {
    core.setFailed((err as Error).message);
  }
}

run();
