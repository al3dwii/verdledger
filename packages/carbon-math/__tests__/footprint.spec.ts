import { describe, expect, it } from "vitest";
import { calculateFootprint } from "../src";
import plan from "../../../testdata/aws-small-plan.json";

describe("footprint maths", () => {
  it("matches golden snapshot", () => {
    const result = calculateFootprint(plan);
    expect(result).toMatchSnapshot();
  });
});
