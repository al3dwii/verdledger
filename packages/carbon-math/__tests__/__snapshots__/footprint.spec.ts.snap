// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`footprint maths > matches golden snapshot 1`] = `
{
  "resourceCount": 1,
}
`;
